### PR TITLE
HTTPCLIENT-1942: Add support for Reactive Streams

### PR DESCRIPTION
**Pull request by rschmitt**
This contribution adds initial support for the [Reactive Streams specification](http://www.reactive-streams.org/). The main part of this change is a pair of non-public classes, the `ReactiveDataProducer` and `ReactiveDataConsumer`, which are reactive adapters for Apache's AsyncDataProducer and AsyncDataConsumer interfaces. Two public classes are built on top of these types:

* `ReactiveEntityProducer`: An `AsyncEntityProducer` implementation backed by `ReactiveDataProducer` that allows a `Publisher<ByteBuffer>` to be streamed as a request body
* `ReactiveResponseConsumer`: An `AsyncResponseConsumer` that exposes the overall HTTP response as a `Message<HttpResponse, Publisher<ByteBuffer>>`, allowing the body to be consumed by a `Subscriber<ByteBuffer>`

These classes are enough to allow the async Apache client to straightforwardly integrate with reactive frameworks like [RxJava](https://github.com/ReactiveX/RxJava), [Akka](https://akka.io/), and [Vert.x](https://vertx.io/).

### CODE REVIEW

File: httpcore5-reactive/pom.xml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      <artifactId>reactive-streams</artifactId>
+      <version>1.0.2</version>
+    </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.reactivex.rxjava2</groupId>
```
Comment (ok2c) [inline httpcore5-reactive/pom.xml]: @rschmitt Can we remove compile dependency on `io.reactivex.rxjava2:rxjava` and make it test only or at least optional?

File: httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveResponseConsumer.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    @Override
+    public void consumeResponse(
+        final HttpResponse response,
+        final EntityDetails entityDetails,
+        final HttpContext httpContext,
+        final FutureCallback<Void> resultCallback
+    ) {
+        this.entityDetails = entityDetails;
+        this.responseCompletion = new BasicFuture<>(resultCallback);
+        this.responseFuture.completed(new Message<HttpResponse, Publisher<ByteBuffer>>(response, reactiveDataConsumer));
```
Comment (ok2c) [inline httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveResponseConsumer.java]: @rschmitt Why not using a simple `Consumer` instead? Do we really need another `Future` here?

File: httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveDataProducer.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            publisher.subscribe(this);
+        }
+
+        final Throwable t = exception.get();
+        final Subscription s = subscription.get();
+        int buffersToReplenish = 0;
+        try {
+            synchronized (buffers) {
+                if (t != null) {
+                    // TODO: We need a reliable way to send RST_STREAM (without subsequent GOAWAY) on HTTP/2.
```
Comment (ok2c) [inline httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveDataProducer.java]: @rschmitt By the way, why not using a subclass of `H2StreamResetException` here?

File: httpcore5-reactive/pom.xml (hunk 1 above)
Comment (rschmitt) [reply httpcore5-reactive/pom.xml]: This was absolutely supposed to be test-only. I'll fix it.

File: httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveDataProducer.java (hunk 3 above)
Comment (rschmitt) [reply httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveDataProducer.java]: That's an interesting option, but what happens if I throw it on HTTP/1.1?

Comment (ok2c) [reply httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveDataProducer.java]: @rschmitt Nothing spectacular. I'll likely be handled like any other I/O error. I am still open to using a special annotation for exceptions, though, if you like that approach better.

File: httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveResponseConsumer.java (hunk 2 above)
Comment (rschmitt) [reply httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveResponseConsumer.java]: I'm not sure what you mean by `Consumer`, because I thought this entire project was targeting JDK7. (RxJava provides a `Consumer<T>` interface, but the Reactive Streams API does not.) Either way, I would argue that we *do* need to return a `Future`, for the same reason that the `#execute` method returns a `Future`: callers need a single thing that they can block on (or decorate with callback logic) to wait for success, failure, or cancellation.

In the integration tests, there is a fairly consistent idiom:

```java
        final ReactiveResponseConsumer consumer = new ReactiveResponseConsumer();
        requester.execute(request, consumer, Timeout.ofSeconds(2), null);
        final Message<HttpResponse, Publisher<ByteBuffer>> response = consumer.getResponseFuture().get();
```

With a `Consumer`-based API, a caller might write:

```java
        final BlockingQueue<Message<HttpResponse, Publisher<ByteBuffer>>> streamingResponse =
            new ArrayBlockingQueue<>(1);
        final ReactiveResponseConsumer consumer =
            new ReactiveResponseConsumer(new Consumer<Message<HttpResponse, Publisher<ByteBuffer>>>() {
                @Override
                public void accept(final Message<HttpResponse, Publisher<ByteBuffer>> httpResponsePublisherMessage) {
                    streamingResponse.offer(httpResponsePublisherMessage);
                }
            });
        requester.execute(request, consumer, Timeout.ofSeconds(2), null);
        final Message<HttpResponse, Publisher<ByteBuffer>> response = streamingResponse.take();
```

However, the above code will block forever if an exception occurs and the response fails to come back. One way to fix this is by setting up a `CompletableFuture` or `BasicFuture` and then completing it from *both* callbacks:

```java
        final CompletableFuture<Message<HttpResponse, Publisher<ByteBuffer>>> streamingResponse =
            new CompletableFuture<>();
        final ReactiveResponseConsumer consumer =
            new ReactiveResponseConsumer(new Consumer<Message<HttpResponse, Publisher<ByteBuffer>>>() {
                @Override
                public void accept(final Message<HttpResponse, Publisher<ByteBuffer>> httpResponsePublisherMessage) {
                    streamingResponse.complete(httpResponsePublisherMessage);
                }
            });
        requester.execute(request, consumer, Timeout.ofSeconds(2), new FutureCallback<Void>() {
            @Override
            public void completed(Void result) {
                // TODO: Can this race with the other callback?
                streamingResponse.completeExceptionally(new RuntimeException("No stream returned"));
            }

            @Override
            public void failed(Exception ex) {
                streamingResponse.completeExceptionally(ex);
            }

            @Override
            public void cancelled() {
                streamingResponse.completeExceptionally(new CancellationException());
            }
        });
        final Message<HttpResponse, Publisher<ByteBuffer>> response = streamingResponse.join();
```

In short, constructing a correctly functioning `Future` for the streaming response is definitely something I want to do on the caller's behalf, rather than asking the caller to construct one manually from lower-level primitives.

File: httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveDataProducer.java (hunk 3 above)
Comment (rschmitt) [reply httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveDataProducer.java]: I'll experiment with `H2StreamResetException`. The suggestion to use annotations is creative, but it may be overkill for what I need, which is probably not much.

Comment (rschmitt) [reply httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveDataProducer.java]: `H2StreamResetException` works perfectly, except for the additional dependency edge from `:httpcore5-reactive` to `:httpcore5-h2`, which I think is acceptable.

Comment (ok2c) [reply httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveDataProducer.java]: @rschmitt I'll refactor that bit once the change-set gets merged.

**Comment by ok2c** [comment]
@rschmitt Committed to master as 1bae9643c912536793e820ac546730871e37155a. Please review and close this PR.

**Comment by rschmitt** [comment]
Thanks for taking the time to review this contribution!
